### DISCUSSION TITLE
Using flashmq_add_task from the plugin

### DISCUSSION THREAD

**Discussion by slavslavov** [halfgaar/FlashMQ#58] [Q&A] [answered]
Will the task be executed in the context of the calling thread? Let's say in `flashmq_plugin_init` I call `flashmq_add_task` to run a task a few seconds later and I capture the `thread_data`. When I get my task called back will it be from the same thread I called `flashmq_add_task` and will `thread_data` be valid at this point?

And if I am running FlashMQ on a 4 core CPU and having 4 threads, I guess my task will be executed 4 times - once per each thread.

**Comment by wiebeytec** [comment, marked as answer]
Yes, the task is placed in the queued tasks list of the current thread's event loop. You can capture `thread_data` and the same thread will perform the task. You can be sure the `thead_data` is not simultanously accessed elsewhere.

And yes, if you have 4 threads, you get four tasks, which you can influence by some kind of static/atomic memory counter, if you want.

[See here BTW](https://github.com/halfgaar/FlashMQ/blob/d0705bf2fa347131f10ad8242bb708a58b1e1ccf/flashmq_plugin.cpp#L183) for what `flashmq_add_task` does.

Side note:

> and will `thread_data` be valid at this point?

By 'valid', you imply that accessing `thread_data` of another thread is invalid. It's fine actually, if properly mutexed, but there should be no reason to do it. The only real time it becomes *invalid* is when you `delete` / `free` `thread_data` outside `flashmq_plugin_allocate_thread_memory()` and `flashmq_plugin_deallocate_thread_memory()`. Edit: and a good reason I should have used C++ style `shared_ptr` instead of C-style `void`. The only reason it's `void`, is for Mosquitto plugin v2 compatibility.
